### PR TITLE
[FSDP] Add Masked importance sampling

### PR DESCRIPTION
**Pull request by zijiexia**
Add masked importance sampling for both token level and sequence level as #1063 .
Results from @GuanxingLu:

Summary:

- Directly use compute_mis_weights func from megatron backend
- Add a pytest file (tests/test_fsdp_mis.py)

run with 4xH200 GPUs (using examples/train_infer_mismatch_helper/run-qwen3-4b-fsdp-mis.sh):
526919906-f205e2da-99f5-4d10-bda4-daad21e88ce9

Unfortunately, the original mismatch of training and rollout engine is quite marginal, so the MIS just has no effect. Thus added a pytest script to test the functionality.

### CODE REVIEW

**Comment by GuanxingLu** [comment]
@PopSoda2002 Hi, could you please review this? Thank you!

File: slime/backends/fsdp_utils/actor.py [hunk 1]
```diff
@@ -655,26 +656,41 @@ def _has_rollout_log_probs(batch) -> bool:
...
+                "cp_group": self.cp_group,
+            }
+
+            if self.args.custom_tis_function_path is not None:
+                tis_func = load_function(self.args.custom_tis_function_path)
+            else:
+                tis_func = vanilla_tis_function_fsdp
+            pg_loss, loss_masks, tis_metrics = tis_func(**tis_kwargs)
+
+        if getattr(self.args, "calculate_per_token_loss", False):
```
Comment (zhaochenyang20) [inline slime/backends/fsdp_utils/actor.py]: A gneral suggestion, avoid using the `getattr` as follows in https://github.com/sgl-project/sglang/pull/14836. It makes the repo unreadable.

File: slime/backends/fsdp_utils/actor.py [hunk 2]
```diff
@@ -721,10 +737,12 @@ def _has_rollout_log_probs(batch) -> bool:
         if self.args.use_opsm:
             reported["opsm_clipfrac"] = opsm_clipfrac
 
-        if self.args.use_tis and tis is not None:
-            reported["tis"] = sum_of_sample_mean(tis, response_lengths, loss_masks).detach()
-            reported["ois"] = sum_of_sample_mean(ois, response_lengths, loss_masks).detach()
-            reported["tis_clipfrac"] = sum_of_sample_mean(tis_clipfrac.float(), response_lengths, loss_masks).detach()
+        if self.args.use_tis and tis_metrics:
+            for k, v in tis_metrics.items():
+                if getattr(self.args, "calculate_per_token_loss", False):
```
Comment (zhaochenyang20) [inline slime/backends/fsdp_utils/actor.py]: avoid using the `getattr`

File: slime/backends/megatron_utils/loss.py [hunk 3]
```diff
@@ -420,7 +420,7 @@ def policy_loss_function(
     Computes current log-probabilities and entropy from model logits, then
     calculates PPO-style clipped policy gradient loss. For GSPO, gathers
     full sequences via context-parallel all-gather before computing per-sample
-    KL. Optionally applies TIS (Temporal Importance Sampling) correction and
+    KL. Optionally applies TIS (Truncated Importance Sampling) correction and
```
Comment (zhaochenyang20) [inline slime/backends/megatron_utils/loss.py:423]: Nice catch 😂

File: tests/test_fsdp_mis.py
Comment (zhaochenyang20) [inline tests/test_fsdp_mis.py:1]: If megatron do not have a test, nor need FSDP.

**Review by zhaochenyang20 — CHANGES_REQUESTED** [review]

File: slime/backends/fsdp_utils/actor.py [hunk 4]
```diff
@@ -803,6 +821,51 @@ def update_weights(self) -> None:  # type: ignore[override]
...
+        elif tis_mode == "mask":
+            mask = (tis >= tis_clip_low) & (tis <= tis_clip_high)
+            tis_clip = tis * mask.float()
+        else:
+            raise ValueError(f"Unsupported tis_mode: {tis_mode}")
+
+        tis_clipfrac = tis_clip != tis
+
+        return tis_clip, tis, tis_clipfrac
+
```
Comment (zhaochenyang20) [inline slime/backends/fsdp_utils/actor.py]: Are you sure that this function should be put under `fsdp_utils`? You can refer to where we put similar function of Megatron.

File: slime/backends/fsdp_utils/actor.py [hunk 5]
```diff
@@ -1103,3 +1171,53 @@ def apply_fsdp2(model, mesh=None, cpu_offload=False, args=None):
...
+    tis_clipfrac = (tis_clip != tis).float()
+
+    metrics = {
+        "tis": tis.clone().detach(),
+        "tis_clipfrac": tis_clipfrac.clone().detach(),
+        "tis_abs": tis_abs.clone().detach(),
+    }
+    pg_loss = pg_loss * tis_clip
+
+    return pg_loss, loss_masks, metrics
```
Comment (zhaochenyang20) [inline slime/backends/fsdp_utils/actor.py]: As I mentioned, maybe these functions should not be put here. We want to keep `actor.py` as clean as possible.

**Review by zhaochenyang20 — CHANGES_REQUESTED** [review]

File: slime/backends/fsdp_utils/actor.py (hunk 5 above)
Comment (GuanxingLu) [reply slime/backends/fsdp_utils/actor.py]: Hi @zijiexia, can we just remove vanilla_tis_function_fsdp (it should not be used as we can specify custom-tis-function-path to use compute_mis_weights in examples/train_infer_mismatch_helper/mis.py) and _compute_tis_weights (it is not used now)?

Comment (zijiexia) [reply slime/backends/fsdp_utils/actor.py]: Hi @GuanxingLu , I've made the following changes:

1. Cleanup the unused functions in `actor.py`.
2. I move `vanilla_tis_function_fsdp` to `ppo_utils.py` as I think we do need it, following the same pattern as Megatron: referring to this https://github.com/THUDM/slime/blob/461fc8a2fe4e7309fa544e7f2287e0343bf2672e/slime/backends/megatron_utils/loss.py#L526
3. I move the `compute_mis_weights_fsdp` to `mis.py`.

I didn't add a new `use-mis` arg as I'm trying to follow the same parameter system in `mis.yaml`. Could you take a look at it and let me know what you think before I mark it back to ready for review? Thanks!

Comment (GuanxingLu) [reply slime/backends/fsdp_utils/actor.py]: Looks good to me!

**Comment by zijiexia** [comment]
Hi @PopSoda2002 @zhaochenyang20 , thanks for the review, I've refactor the code accordingly:

1. Delete the test script.
2. Remove the unused functions (sorry!) and cleanup the `actor.py`
3. I didn't add an additional `use-mis` args since I was trying to follow the same pattern as https://github.com/THUDM/slime/pull/429. Please let me know what you think, I'll further refactor accordingly.

Thank you!

**Comment by zhaochenyang20** [comment]
thanks! Sorry for the late reply. Let me and Huapeng review this @PopSoda2002

**Comment by zhaochenyang20** [comment]
Nice done Zijie! @zijiexia

**Comment by PopSoda2002** [comment]
> Nice done Zijie! @zijiexia

I think @GuanxingLu does important contribute to this PR also cc @zhaochenyang20 😂

**Comment by zijiexia** [comment]
> Nice done Zijie! @zijiexia

@GuanxingLu start this before I join so most credit should goes to him

**Comment by GuanxingLu** [comment]
Appreciate it, we all made a lot, happy to contribute!
